keg_auth/model/__init__.py

```diff
@@ -0,0 +1,521 @@
+            cls._query_bundle_permissions(),
+            cls._query_group_permissions()
+        )
+
+    @classmethod
+    def _query_direct_permissions(cls):
+        perm_cls = entity_registry.registry.permission_cls
+        return db.session.query(
+            cls.id.label('user_id'),
+            perm_cls.id.label('perm_id'),
+        ).select_from(
+            cls
+        ).join(
+            cls.permissions
+        )
+
+    @classmethod
+    def _query_bundle_permissions(cls):
+        perm_cls = entity_registry.registry.permission_cls
+        bundle_cls = entity_registry.registry.bundle_cls
+        return db.session.query(
+            cls.id.label('user_id'),
+            perm_cls.id.label('perm_id'),
+        ).select_from(
+            cls
+        ).join(
+            cls.bundles
+        ).join(
+            bundle_cls.permissions
+        )
+
+    @classmethod
+    def _query_group_permissions(cls):
+        group_cls = entity_registry.registry.group_cls
+
+        group_mapping = group_cls._query_permission_mapping().alias('group_permissions_mapping')
+        return db.session.query(
+            cls.id.label('user_id'),
+            group_mapping.c.perm_id.label('perm_id')
+        ).select_from(
+            cls
+        ).join(
+            cls.groups
+        ).join(
+            group_mapping,
+            group_mapping.c.group_id == group_cls.id
+        )
+
+
+class UserEmailMixin(object):
+    # Assume the user will need to verify their email address before they become active.
+    is_verified = sa.Column(sa.Boolean, nullable=False, default=False,
+                            server_default=sa.text('false'))
+    email = sa.Column(EmailType, nullable=False, unique=True)
+    token = sa.Column(PasswordType(onload=_create_cryptcontext_kwargs))
+    token_created_utc = sa.Column(ArrowType)
+
+    @hybrid_property
+    def is_active(self):
+        return self.is_verified and self.is_enabled
+
+    @is_active.expression
+    def is_active(cls):
+        return sa_sql.and_(cls.is_verified == sa.true(), cls.is_enabled == sa.true())
+
+    @classmethod
+    def testing_create(cls, **kwargs):
+        # Most tests will want an active user by default, which is the opposite of what we want in
+        # production, so swap that logic.
+        kwargs.setdefault('is_verified', True)
+
+        user = super(UserEmailMixin, cls).testing_create(**kwargs)
+        return user
+
+    def token_verify(self, token):
+        # If a token isn't set, it's can't be verified.
+        if token is None or self.token is None or self.token_created_utc is None:
+            return False
+
+        # The token is invalid if it has expired.
+        expire_mins = flask.current_app.config['KEGAUTH_TOKEN_EXPIRE_MINS']
+        expire_at = self.token_created_utc.shift(minutes=expire_mins)
+        if arrow.get() >= expire_at:
+            return False
+
+        return self.token == token
+
+    @might_commit
+    def token_generate(self):
```

Comment:
This is not a sufficiently secure method of generating a token. It leaves the possibility for replay attacks when a developer messes something up or forgets to do a step. 

A secure token is one the encodes the context of the user when the token is generated. When verifying the token, that context is validated as still "current" and then allows password reset.

Relevant context would include, email address, current password hash, enabled flag / verified flags, last update time, and then the expiration of the token. Those values should be string coerced, signed, and that value (or some substring) should be used as the token.


For example, 
- if a user asks for a password reset, then, before they reset it they are disabled, they should not be able to reset the password.
- If they request a password, but an admin has reset it, they reset link should no longer work (regardless of what is in the database). In this situation especially, you now have to worry about changing the password in multiple places and ensuring that you clear the token properly, etc.
- I would venture to guess there are other ways this could break, but I will leave it alone for now.